learn/sherif/main_test/src/main/java/frc/robot/RobotContainer.java

```diff
@@ -26,26 +28,37 @@
  */
 public class RobotContainer {
     // The robot's subsystems and commands are defined here...
-    private Factory m_f = new Factory();
-    private DriveTrain m_drivetrain;
-    private Joystick m_controller;
+    public Factory factory = new Factory();
+    private DriveTrain m_drivetrain = null;
+    private Joystick m_controller = null;
+    private IntakeControl m_intake = null;
 
     /**
      * The container for the robot.  Contains subsystems, OI devices, and commands.
      */
     public RobotContainer() {
         // Configure the button bindings
+        this.m_drivetrain = new DriveTrain(factory);
+        this.m_controller = new Joystick(0);
         configureButtonBindings();
     }
 
     // Allows for a (mock) controller to be injected
-    public RobotContainer(Joystick controller, DriveTrain train) {
-        this.m_controller = controller;
-        this.m_drivetrain = train;
+    public RobotContainer(RobotInjection injection) {
+        if (injection.joystick != null) {
+            this.m_controller = injection.joystick;
+        }
+        if (injection.driveTrain != null) {
+            this.m_drivetrain = injection.driveTrain;
+        }
+        if (injection.intakeControl != null) {
+            this.m_intake = injection.intakeControl;
+        }
```

Comment:
I generally prefer initializing variables in exactly one place to avoid bugs when a variable is added and one constructor gets updated and not the other.  In your code you are initializing the `m_intake` variable her but not in the default constructor. 

I think we should probably have all these variables always initialized to non-null values to avoid having to check for nulls in other parts of the code.

In Java you can call one constructor from another.

There are a couple solutions.  One would be to write:
```
public RobotContainer() {
  this(new RobotInjection())
}
public RobotContainer(RobotInjection injection) {
  this.m_controller = injection.joystock ? injection.joystick : new Joystick(0);
  this.m_drivetrain = injection.driveTrain ? injection.driveTrain : new DriveTrain(factory);
  this.m_intake = injection.intakeControl ? injection.intakeControl : /* WHAT GOES HERE? */
}
```

Reply:
That code would be better but I think an issue that comes up with mainly testing is what if you don't want to inject everything. Say in the intake test you want to test controller commands only with `Intake` and not including `Drive`, instead of being forced to input injections of irrelevant subsystems. I'll see if I can find some solution and comment here again